xapian-applications/omega/worker.cc

```diff
@@ -169,22 +171,19 @@ Worker::extract(const std::string& filename,
 		    !read_string(sockt, strpage)) {
 		    break;
 		}
-		if (strpage.empty() ||
-		    strpage.find_first_not_of("0123456789") != string::npos) {
+		if (!parse_signed(strpage.c_str(), pages) || pages < 0)
 		    pages = 0;
-		} else {
-		    pages = stoi(strpage);
-		}
 		return true;
 	    case MSG_NON_FATAL_ERROR:
-		if (strstate.length() > 1 && verbose) {
-		    cerr << (strstate.c_str() + 1) << endl;
+		if (strstate.length() > 1) {
+		    error = (strstate.c_str() + 1);
```

Comment:
This effectively converts the C++ string to a C string and back, which is probably better avoided - the conversion back will take the length by searching for the first zero byte, which means a string with embedded zero bytes will be truncated, and is also a little less efficient than just taking the known length of the C++ string.  The truncation issue isn't likely to be a problem in practice in this situation, but still seems a bit unclean.

You can assign a substring of a C++ string like this:

```
error.assign(strstate, 1, string::npos)
```

(The third argument isn't needed from C++14 on, but currently we aim to support C++11).

Reply:
Okay, I am going to change it.